### ISSUE TITLE
Add "echo data" in the quote response and transfer request

### ISSUE THREAD

**Issue by adrianhopebailie** [mojaloop/mojaloop-specification#14]
If #13 is implemented to align the API with ILPv4 it will be necessary to define a data element in the callback of the `/quote` API to carry data sent by the payee that it expects to be unchanged and returned in the subsequent `/transfer` API call. (Currently payee systems will use the encoded ILP packet)

This field will be used by the payee to generate the fulfilment and condition so it must be returned unchanged to the payee in order to fulfil the transfer. To satisfy this requirement it is recommended that the field is base64 encoded.

The content of the field is at the discretion of the payee and MAY contain data (such as the quote and transaction details) that it can use to ensure the fulfilment and condition are cryptographically bound to the transaction.

**Comment by ehenrka** [comment]
This change request is very abstract in its current state. It is difficult for me to comment on this change without any more detail. Can you please add some more detail regarding how you would like this to look like in the API? 

Why does the field need to be base64 encoded if it is just one field? If there were several elements then I could understand the need for base64 encoding to preserve structure of the data in JSON, but for just one element there should be no need to preserve structure. You cannot guarantee that an element is unchanged just by having it base64 encoded. To guarantee that the element is unchanged you should have end to end signing of the element.

**Comment by ehenrka** [comment]
Thank you @adrianhopebailie for providing additional details.

There are other fields which could have more of an impact if the data was unintendedly encoded and decoded improperly by JSON parsers along the way. I'm having troubles justifying the use of base64 for just this element. This would be the only element in the API that uses BinaryString of undefined length as the IlpPacket would now be removed (IlpCondition and IlpFulfilment uses BinaryString32).

Otherwise I'm fine with the change.

**Comment by adrianhopebailie** [comment]
@ehenrka - given the requirements; an opaque data element that must be identical (byte for byte) when sent in the quote callback and the transfer request, I think this SHOULD be sent as binary data.

Can you suggest a better way to do this than base64 encoding?

**Comment by ehenrka** [comment]
@adrianhopebailie Our existing signature functionality which is based on JWS also requires identical data (byte for byte). This means that if the data which is signature protected is modified on its way the signature validation will fail, thus there is already the same requirement (byte for byte identical) for all data.

To have the data opaque, an FSP can encode it in whatever way they like (e.g. base64 or some other encoding). My suggestion would be that we recommend base64 encoding in the spec, but keep it open to the FSP which encoding that they would like as this data should only be used by the Payee FSP anyways. We can still keep a similar regex, e.g. `^[A-Za-z0-9-_=\/,.:~]+$` (existing `BinaryString` is `^[A-Za-z0-9-_]+[=]{0,2}$)`. This would leave it up to the implementation which base64 variant to use and if the ending padding characters should be present or not, without us describing it in the spec (the current text in the spec led to some problems as one implementer interpreted it incorrectly).

**Comment by adrianhopebailie** [comment]
@ehenrka what about making this just UTF8 text with a safe size limit?

Then implementations can put whatever they want. If they want to use binary data they can base64 encode it or send it as hex.

**Comment by ehenrka** [comment]
@adrianhopebailie Sounds good to me!

**Comment by MichaelJBRichards** [comment]
I agree. It seems to me that the data is for the sole use of the payee, and it is up to the payee to decide _both_ whether or not to include an encoded version of the data it wants to check, _and_ which elements of the transfer should be included in the check it performs (whether or not an encoded version is passed.) A given scheme might, of course, decide to mandate the form of this check; but, for me, that's not the same as saying that the API definition needs to codify it.

I assume also that the response from the payee to a quote request will need to include the Transaction object, since it's the payee who defines the Transaction object. Is that correct?
